File: leetcode_C++/p918.cpp
/*
@Author: mrxhar
@Date:   2023-09-10 10:44:17
@Email:  1337904145@qq.com
@Blog:   https://leetcode.cn/problems/maximum-sum-circular-subarray/?envType=daily-question&envId=2023-09-10
*/
#include <deque>
#include <vector>

using namespace std;

class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int total = 0, maxSum = nums[0], curMax = 0, minSum = nums[0], curMin = 0;
        for (int& num : nums) {
            curMax = max(curMax + num, num);
            maxSum = max(maxSum, curMax);
            curMin = min(curMin + num, num);
            minSum = min(minSum, curMin);
            total += num;
        }
        return maxSum > 0 ? max(maxSum, total - minSum) : maxSum;
    }
};

File: leetcode_py/p918.py
from typing import List


class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        # 直接使用 最大最小
        n = len(nums)
        maxSum = nums[0]
        minSum = nums[0]
        curMax = curMin = 0
        total = 0
        for num in nums:
            total += num
            curMax = max(0,curMax) + num
            curMin = min(0,curMin) + num

            maxSum = max(maxSum ,curMax)
            minSum = min(minSum,curMin)
        return maxSum if maxSum > 0 else total - minSum
            